### ISSUE TITLE
Migrate remaining foreign keys pointing to `builds (id)` from `integer` to `bigint`

### ISSUE THREAD

**Issue by vito** [concourse/concourse#6199]
[This migration](https://gist.github.com/vito/05a210b79082906280baff013e972428) should do, except for one problem: `build_events` will take _**forever**_ to migrate naively like this.

Consider adding a new column instead and incrementally migrating it. See [this post](https://techcommunity.microsoft.com/t5/azure-database-for-postgresql/postgres-tips-how-to-convert-2-billion-rows-to-bigint-with-citus/ba-p/1490128) as an example. Ideally we can have this happen transparently through some component that runs in the background so end-users don't have to do anything.

**Comment by vito** [comment]
Going to pick this up and get it over with so v6.7.0 doesn't have a ticking time-bomb in it.

Already investigated this last night and have a working proof of concept - gonna write some tests for it though.

One thing I'm considering is having the batch process run only periodically so it doesn't hammer the database with writes all the time. It'll migrate the most recent builds first, so it should get the most useful ones done pretty quickly, and the rest can happen gradually.

Since operators may want to tune this, I anticipate there being two flags so far:

```
--build-bigint-incremental-migration-batch-size 1000
--build-bigint-incremental-migration-wait 10s
```

Not sure what the defaults will be, but the above would migrate 1000 builds at a time, waiting 10 seconds in between batches.

Note that 1000 builds doesn't really say how many *events* they have - it's kinda hard to be precise around that without making the queries slower/more complicated, since events don't have their own globally unique autoincrementing ID - they have a two-value primary key of (`build_id`, `event_id`), where `event_id` starts at 1 for each build.

I'm running my experiments against 100,000 builds with an even 500 events for each build, which I generated with this query:

```sql
INSERT INTO build_events (build_id, type, payload, event_id, version)
SELECT build.id, 'log', '{"origin":{"id":"some-plan-id","source":"stderr"},"payload":"hello","time":123}', event.id, '1.0'
FROM generate_series(1, 100000) AS build (id), generate_series(1, 500) AS event (id);
```

Which led me to discovering that build events don't have a foreign key reference for the `build_id` column. 🤔 This probably means that resource checking builds are leaking their build events forever. This is probably so we can just truncate/drop the whole table partition when a pipeline is deleted - but maybe we could just truncate it first so the cascade doesn't fire? Maybe we need a new garbage collector to remove build events for deleted builds? ...might be challenging to even find which ones to remove efficiently though. Hmm....

... anyway, 500 events seems somewhat realistic based on some randomly selected data from `ci`:

```sql
WITH events_per_build AS (
    SELECT max(event_id) as events
    FROM pipeline_build_events_6
    GROUP BY build_id
    LIMIT 1000
)
SELECT avg(events) FROM events_per_build;
-- => 582
```

Obviously there's a lot of variance here. It would be nice if I could run against a live database, but GCP's CloudSQL doesn't seem to really have a way to easily do that. (The backup system is *very* MVP, and only supports restoring.)

**Comment by aoldershaw** [comment]
One slightly concerning thing about a using a fixed build limit instead of a fixed event limit is that sometimes builds can spin forever and produce millions of log events. #6080 reports the DB hanging as it tries to delete build events. Granted, we don't have a batch size for deletion of build events (we seem to have a batch size of 1000 builds, but that's just for paginating the builds - we could end up trying to clear out >1000 builds at a time).

So, even if most of the time builds are relatively small, there could be some really huge builds caused by e.g. errant tasks, and those builds *may* cause issues. Could we instead `ORDER BY build_id DESC, event_id DESC` and `LIMIT 50000`?

**Comment by vito** [comment]
@aoldershaw Good point! I think I found a way to support this. Postgres supports comparing `(int, int)`, and we have a primary key on `(build_id, event_id)` - so it's actually not too bad to just walk over it using a combination of `LIMIT`/`OFFSET [batchSize]` to figure out the boundary for each update.

So this way the batch size can be configured as an exact number of events to update at a time, which will have *much* more predictable performance.

**Comment by aoldershaw** [comment]
Not sure I fully see it - does that work if we go in descending order through the builds? We can offset by how ever many events we have processed, but as new events come in, would we need to shift that offset somehow?

**Comment by vito** [comment]
@aoldershaw The migration does a little column renaming dance so that we end up with `build_id` as the bigint column and `build_id_old` as the old column (with all the old data). So all new events will be written as `bigint`s even if they're written by an "old" `web` node during the upgrade, and we'll only have to worry about migrating the fixed amount of data that was written before the migration ran. (I tested it out and renaming columns is fast; the number of rows doesn't seem to matter.)

I've got tests written for this and it seems to all be working - will push some code up soon!

I'm trying to decide when/how to run this thing now. 🤔 Quick-and-dirty solution is to just add an ifrit runner that does the batch migration and just sits around when it's done. This raises a few questions though:

* When do we remove it from the set of runners? Leaving it there forever kind of feels like code cruft. It'd be nice if it was out of sight like the other migrations.
* How do we make sure only one `web` node runs it? Guess I'll add a lock? Maybe I could just make this a component so it gets locking for free?
* How does the runner know when it's done? Right now it just catches an `undefined_column` error to know it already migrated everything and cleaned up the schema. It'd be nice if we kept track of them just like we keep track of normal migrations and ran it only once.

These are all things that are solved by 'normal' migrations, so I'm kind of tempted to work this into a feature for [Voyager](https://github.com/concourse/voyager) to support incremental migrations. But I just remembered we never actually switched to it (there's still a WIP branch), so that's a whole detour I'll probably avoid for now.

(Also: gonna see how this performs with a bunch of data.)
